File: react-kanban/src/types/state/board.ts
import { ColumnType } from "../serialize";
import { createId } from "../utility";
import {
  canColumnAcceptTaskFrom,
  columnTypes,
  createColumn,
  type Column,
} from "./column";
import type { Task, TaskId } from "./task";

export type BoardId = string;

export type Board = {
  id: BoardId;
  title: string;
  columns: Column[];
};

type MoveTaskContext = {
  sourceColumn: Column;
  targetColumn: Column;
  movingTask: Task;
};

export function createBoard(title: string): Board {
  return {
    id: createId(),
    title,
    columns: columnTypes.map((columnType) => createColumn(columnType, null)),
  };
}

export function addTaskToBoardColumn(
  board: Board,
  columnType: ColumnType,
  task: Task,
): Board {
  return {
    ...board,
    columns: board.columns.map((column) =>
      column.columnType === columnType
        ? { ...column, tasks: [...column.tasks, task] }
        : column,
    ),
  };
}

export function updateBoardTask(board: Board, taskId: TaskId, nextTask: Task): Board {
  let changed = false;
  const columns = board.columns.map((column) => {
    let columnChanged = false;
    const tasks = column.tasks.map((task) => {
      if (task.id !== taskId) {
        return task;
      }

      columnChanged = true;
      changed = true;
      return nextTask;
    });

    return columnChanged ? { ...column, tasks } : column;
  });

  return changed ? { ...board, columns } : board;
}

export function deleteBoardTask(board: Board, taskId: TaskId): Board {
  let changed = false;
  const columns = board.columns.map((column) => {
    const tasks = column.tasks.filter((task) => task.id !== taskId);

    if (tasks.length === column.tasks.length) {
      return column;
    }

    changed = true;
    return { ...column, tasks };
  });

  return changed ? { ...board, columns } : board;
}

export function canMoveBoardTask(
  board: Board,
  taskId: TaskId,
  toColumnType: ColumnType,
  beforeTaskId: TaskId | null,
): boolean {
  return getMoveTaskContext(board, taskId, toColumnType, beforeTaskId) !== null;
}

export function moveBoardTask(
  board: Board,
  taskId: TaskId,
  toColumnType: ColumnType,
  beforeTaskId: TaskId | null,
): Board {
  const moveContext = getMoveTaskContext(
    board,
    taskId,
    toColumnType,
    beforeTaskId,
  );
  if (!moveContext) {
    return board;
  }

  const { sourceColumn, targetColumn, movingTask } = moveContext;

  const columns = board.columns.map((column) => {
    if (column.id !== sourceColumn.id && column.id !== targetColumn.id) {
      return column;
    }

    let tasks = column.tasks;

    if (column.id === sourceColumn.id) {
      tasks = tasks.filter((task) => task.id !== taskId);
    }

    if (column.id === targetColumn.id) {
      const insertIndex = beforeTaskId
        ? tasks.findIndex((task) => task.id === beforeTaskId)
        : -1;
      const normalizedInsertIndex =
        insertIndex >= 0 ? insertIndex : tasks.length;

      tasks = [
        ...tasks.slice(0, normalizedInsertIndex),
        movingTask,
        ...tasks.slice(normalizedInsertIndex),
      ];
    }

    return { ...column, tasks };
  });

  return { ...board, columns };
}

function getMoveTaskContext(
  board: Board,
  taskId: TaskId,
  toColumnType: ColumnType,
  beforeTaskId: TaskId | null,
): MoveTaskContext | null {
  if (beforeTaskId === taskId) {
    return null;
  }

  const sourceColumn = board.columns.find((column) =>
    column.tasks.some((task) => task.id === taskId),
  );
  const targetColumn = board.columns.find(
    (column) => column.columnType === toColumnType,
  );

  if (!sourceColumn || !targetColumn) {
    return null;
  }

  if (!canColumnAcceptTaskFrom(targetColumn, sourceColumn.columnType)) {
    return null;
  }

  const movingTask = sourceColumn.tasks.find((task) => task.id === taskId);
  if (!movingTask) {
    return null;
  }

  return {
    sourceColumn,
    targetColumn,
    movingTask,
  };
}

File: leptos-kanban/src/types/state/board.rs
use leptos::prelude::*;

use crate::types::{
    serialize::{Board, BoardId, ColumnType, TaskId},
    state::{column::ColumnState, task::TaskState},
};

#[derive(Clone, Copy, Debug)]
pub struct BoardState {
    pub id: BoardId,
    pub title: RwSignal<String>,
    pub columns: RwSignal<Vec<ColumnState>>,
}

impl BoardState {
    pub fn from_board(board: &Board) -> Self {
        board.into()
    }

    pub fn add_task_to_column(&self, column_type: &ColumnType, task: TaskState) {
        let column = self.columns.with(|columns| {
            columns
                .iter()
                .copied()
                .find(|column| &column.column_type == column_type)
        });

        if let Some(column) = column {
            column.tasks.update(|tasks| tasks.push(task));
        }
    }

    pub fn delete_task(&self, task_id: &TaskId) {
        let target_column = self.columns.with(|columns| {
            columns.iter().copied().find(|column| {
                column
                    .tasks
                    .with(|tasks| tasks.iter().any(|task| &task.id == task_id))
            })
        });

        if let Some(column) = target_column {
            column
                .tasks
                .update(|tasks| tasks.retain(|task| &task.id != task_id));
        }
    }

    pub fn move_task(
        &self,
        task_id: &TaskId,
        to_column_type: &ColumnType,
        before_task_id: Option<TaskId>,
    ) -> bool {
        if before_task_id.as_ref() == Some(task_id) {
            return false;
        }

        let source_column = self.find_column_containing_task(task_id);
        let target_column = self.find_column_by_type(to_column_type);

        let (Some(source_column), Some(target_column)) = (source_column, target_column) else {
            return false;
        };

        if !target_column.can_accept_task_from(source_column.column_type) {
            return false;
        }

        if source_column.id == target_column.id {
            return source_column.move_task(task_id, before_task_id);
        }

        let Some(task) = source_column.remove_task(task_id) else {
            return false;
        };

        target_column.insert_task(task, before_task_id);
        true
    }

    fn find_column_by_type(&self, column_type: &ColumnType) -> Option<ColumnState> {
        self.columns.with(|columns| {
            columns
                .iter()
                .copied()
                .find(|column| &column.column_type == column_type)
        })
    }

    fn find_column_containing_task(&self, task_id: &TaskId) -> Option<ColumnState> {
        self.columns.with(|columns| {
            columns
                .iter()
                .copied()
                .find(|column| column.has_task(task_id))
        })
    }
}

impl From<&Board> for BoardState {
    fn from(board: &Board) -> Self {
        Self {
            id: board.id(),
            title: RwSignal::new(board.title().to_string()),
            columns: RwSignal::new(board.columns().iter().map(ColumnState::from).collect()),
        }
    }
}
